### PR TITLE
Fix the conversion of Prod values in the native compiler.

### LINKED ISSUE #16645: Inconsistent handling of products in the native machine
```coq
Goal False.
Proof.
assert (H: forall x:unit, False).
{ native_cast_no_check (fun x:unit => I). }
exact (H tt).
Qed.
```

### PR DESCRIPTION
**Pull request by ppedrot**
Fixes #16645: Inconsistent handling of products in the native machine.

- [X] Added / updated **test-suite**.
- [x] Added **changelog**.
- [x] Added / updated **documentation**.

### CODE REVIEW

**Comment by ppedrot** [comment]
cc @silene

File: dev/doc/critical-bugs [hunk 1]
```diff
@@ -386,6 +386,16 @@ Conversion machines
   risk: none, unless mixing open terms and primitive floats inside primitive
   arrays; critical otherwise
 
+  component: "native" conversion machine (translation to OCaml which compiles to native code)
+  summary: conversion of Prod / Prod values was comparing the wrong components
+  introduced: V8.5
+  impacted released versions: V8.5-V8.16.0
```
Comment (Blaisorblade) [inline dev/doc/critical-bugs:392]: _Might_ be worth clarifying that `native_compute` might not be enabled by default and point to https://coq.discourse.group/t/about-the-native-compiler-flag-and-the-coq-native-opam-package-to-be-released-before-coq-8-13-beta1/1147 or similar? I just confirmed I don't have `coq-native` installed and the example doesn't go through for me.

Comment (proux01) [reply dev/doc/critical-bugs:392]: Well, this is only since 8.13, before that the default was "ondemand", so the explanation might get a tad lengthy.

**Review by maximedenes — APPROVED** [review]
Ouch! We should really write unit tests...

**Comment by ppedrot** [comment]
The author of 2cceef0e3cab18b1dcc28bf1c8ce6b4723cd3d9a should probably have realized that there were something very wrong going on, but well...

**Comment by ppedrot** [comment]
Any takers?

**Comment by proux01** [comment]
@coqbot run ci now

**Comment by proux01** [comment]
@coqbot run full ci

File: dev/doc/critical-bugs (hunk 1 above)
Comment (Blaisorblade) [reply dev/doc/critical-bugs:392]: "when built with native_compute enabled" is the only new info, the rest can be handled by links to existing docs.

**Comment by proux01** [comment]
@coqbot run full ci

File: dev/doc/critical-bugs (hunk 1 above)
Comment (proux01) [reply dev/doc/critical-bugs]: @Blaisorblade ok, let's do that

**Comment by proux01** [comment]
@coqbot merge now

**Pull request merged**